src/extra/sprite.cpp

```diff
@@ -0,0 +1,453 @@
+#include "./sprite.hpp"
+
+namespace chainblocks {
+namespace Sprite {
+
+struct Sheet {
+  static inline Type ObjType{
+      {CBType::Object, {.object = {.vendorId = CoreCC, .typeId = SheetCC}}}};
+  static inline Type VarType = Type::VariableOf(ObjType);
+
+  static inline ObjectVar<Sheet> Var{"Sprite-Sheet", CoreCC, SheetCC};
+
+  static CBTypesInfo inputTypes() { return CoreInfo::NoneType; }
+
+  static CBTypesInfo outputTypes() { return Sheet::ObjType; }
+
+  static CBParametersInfo parameters() { return _params; }
+
+  void setParam(int index, const CBVar &value) {
+    switch (index) {
+    case 0:
+      _atlas = value;
+      break;
+
+    default:
+      throw CBException("Parameter out of range.");
+    }
+  }
+
+  CBVar getParam(int index) {
+    switch (index) {
+    case 0:
+      return _atlas;
+    }
+    throw CBException("Parameter out of range.");
+  }
+
+  void cleanup() { _atlas.cleanup(); }
+
+  void warmup(CBContext *context) { _atlas.warmup(context); }
+
+  CBVar activate(CBContext *context, const CBVar &input) {
+    auto table = _atlas.get().payload.tableValue;
+    auto api = table.api;
+
+    auto size = api->tableAt(table, "size")->payload.seqValue;
+    this->width = size.elements[0].payload.intValue;
+    this->height = size.elements[1].payload.intValue;
+
+    this->premultiply = api->tableAt(table, "pma")->payload.boolValue;
+
+    auto regions = api->tableAt(table, "regions")->payload.seqValue;
+    this->regions.reserve(regions.len);
+    for (uint32_t i = 0; i < regions.len; ++i) {
+      auto table = regions.elements[i].payload.tableValue;
+      auto api = table.api;
+
+      auto region = Region{};
+
+      auto bounds = api->tableAt(table, "bounds")->payload.seqValue;
+      region.width = bounds.elements[2].payload.intValue;
+      region.height = bounds.elements[3].payload.intValue;
+
+      if (api->tableContains(table, "uvs")) {
+        auto uvs = api->tableAt(table, "uvs")->payload.seqValue;
+        for (uint32_t j = 0; j < uvs.len && j < 4; ++j) {
+          region.uvs[j] = (float)uvs.elements[j].payload.floatValue;
+        }
+      } else {
+        // calculate UVs
+        double x = bounds.elements[0].payload.intValue;
+        double y = bounds.elements[1].payload.intValue;
+        double u = x / (double)this->width;
+        double v = y / (double)this->height;
+        bool rotate = api->tableAt(table, "rotate")->payload.boolValue;

```

Comment:
Yeah this is a rotation in the atlas texture. It only matters when calculating UVs. It is just in case those were not already included in the .edn file.

Reply:
shouldn't it also be rotated back when rendering? (x1=y0, y1=-x0)
currently it only swaps the width/height from the bounds

Reply:
That's only what's needed.